jvm-libs/test-utils/src/main/kotlin/maru/test/genesis/BesuGenesisFactory.kt

```diff
@@ -0,0 +1,185 @@
+import maru.consensus.DifficultyAwareQbftConfig
+import maru.consensus.ElFork
+import maru.consensus.ForksSchedule
+import org.hyperledger.besu.tests.acceptance.dsl.node.configuration.genesis.GenesisConfigurationFactory
+
+class BesuGenesisFactory(
+  val genesisTemplate: String = genesisTemplateLondonWithoutConsensus,
+  val cliqueBlockTimeSeconds: UInt = 1u,
+  val cliqueEmptyBlocks: Boolean = true,
+) {
+  private var forkSchedule: ForksSchedule? = null
+
+  // we can cache the genesis if for the same fork schedule
+  // which does not change per cluster (for now at least)
+  // building the genesis for many nodes repeatedly is wasteful
+  private var cachedGenesis: String? = null
+
+  fun setForkSchedule(value: ForksSchedule) {
+    this.forkSchedule = value
+    // invalidate cached genesis when fork schedule changes
+    cachedGenesis = null
+  }
+
+  fun create(): String {
+    if (this.forkSchedule == null) {
+      throw IllegalStateException("forkSchedule must be initialized before calling creating genesis")
+    }
+
+    return if (cachedGenesis != null) {
+      return cachedGenesis!!
+    } else {
+      cachedGenesis =
+        createGenesisWithClique(
+          genesisTemplate,
+          cliqueBlockTimeSeconds,
+          cliqueEmptyBlocks,
+          forkSchedule!!,
+        )
+      println("new genesis created: $cachedGenesis")
+      cachedGenesis!!
+    }
+  }
+
+  companion object {
+    val genesisTemplateLondonWithoutConsensus: String =
+      GenesisConfigurationFactory.readGenesisFile("/besu-genesis-template.json")
+    private val jsonObjectMapper: ObjectMapper = jacksonObjectMapper()
+
+    fun createGenesisWithClique(
+      genesisTemplate: String = genesisTemplateLondonWithoutConsensus,
+      chainId: ULong,
+      cliqueBlockTimeSeconds: UInt,
+      cliqueEmptyBlocks: Boolean = true,
+    ): String {
+      var updatedGenesis = genesisTemplate
+      updatedGenesis = setGenesisConfigProperty(updatedGenesis, "chainId", chainId)
+      updatedGenesis = setGenesisCliqueOptions(updatedGenesis, cliqueBlockTimeSeconds, cliqueEmptyBlocks)
+      return updatedGenesis
+    }
+
+    fun createGenesisWithClique(
+      genesisTemplate: String = genesisTemplateLondonWithoutConsensus,
+      cliqueBlockTimeSeconds: UInt,
+      cliqueEmptyBlocks: Boolean = true,
+      forks: ForksSchedule,
+    ): String {
+      var terminalTotalDifficulty: ULong? =
+        forks.forks
+          .firstOrNull { it.configuration is DifficultyAwareQbftConfig }
+          ?.let {
+            (it.configuration as DifficultyAwareQbftConfig).terminalTotalDifficulty
+          } ?: 0UL
+
+      var shanghaiTimestamp: ULong? = null
+      var cancunTimestamp: ULong? = null
+      var pragueTimestamp: ULong? = null
+      forks.forks.forEach { forkSpec ->
+        when (forkSpec.configuration.fork.elFork) {
+          ElFork.Paris -> {} // nothing to do, terminalTotalDifficulty already set
```

Comment:
I think it's not necessarily set. But if it's not set and the first fork is Paris, it should work fine

Reply:
Nevermind, that was irrelevant